### ISSUE TITLE
Maintenance task CPU load

### ISSUE THREAD

**Issue by bobobo1618** [timescale/promscale#596]
I'm running a Git build, commit b4424436f02b200d696c50fdc3d3e9faa6b66d24.

I have two hopefully-related problems:
- There are _always_ two processes running on my system, each consuming an entire CPU core: `postgres: User-Defined Action [1000]` and `postgres: User-Defined Action [1000]`.
- When I run `CALL prom_api.execute_maintenance();`, (as is recommended by the installation docs) it takes 73 minutes to complete.

Hopefully helpful information:

```
prometheus=# SELECT * FROM timescaledb_information.job_stats;
 hypertable_schema | hypertable_name | job_id |      last_run_started_at      |    last_successful_finish     | last_run_status | job_status | last_run_duration |          next_start           | total_runs | total_successes | total_failures 
-------------------+-----------------+--------+-------------------------------+-------------------------------+-----------------+------------+-------------------+-------------------------------+------------+-----------------+----------------
                   |                 |      1 | 2021-04-20 02:44:01.57687+02  | 2021-04-20 02:44:10.940073+02 | Success         | Scheduled  | 00:00:09.363203   | 2021-04-21 02:44:10.940073+02 |         31 |              29 |              2
                   |                 |   1000 | 2021-04-20 19:06:32.465221+02 | 2021-04-20 18:36:32.461375+02 |                 | Running    |                   | -infinity                     |        346 |             204 |            129
                   |                 |   1001 | 2021-04-20 19:06:08.151143+02 | 2021-04-20 18:36:08.147095+02 |                 | Running    |                   | -infinity                     |        352 |             228 |            123
(3 rows)

prometheus=# SELECT * FROM timescaledb_information.jobs;
 job_id |      application_name      | schedule_interval | max_runtime | max_retries | retry_period |      proc_schema      |        proc_name        |  owner   | scheduled | config |          next_start           | hypertable_schema | hypertable_name
 
--------+----------------------------+-------------------+-------------+-------------+--------------+-----------------------+-------------------------+----------+-----------+--------+-------------------------------+-------------------+----------------
-
      1 | Telemetry Reporter [1]     | 24:00:00          | 00:01:40    |          -1 | 01:00:00     | _timescaledb_internal | policy_telemetry        | postgres | t         |        | 2021-04-21 02:44:10.940073+02 |                   | 
   1000 | User-Defined Action [1000] | 00:30:00          | 00:00:00    |          -1 | 00:05:00     | _prom_catalog         | execute_maintenance_job | postgres | t         |        | -infinity                     |                   | 
   1001 | User-Defined Action [1001] | 00:30:00          | 00:00:00    |          -1 | 00:05:00     | _prom_catalog         | execute_maintenance_job | postgres | t         |        | -infinity                     |                   | 
(3 rows)
```

From some spelunking through the codebase, it _looks_ like the two jobs that Promscale apparently created itself are both `execute_maintenance_job`. From [here](https://github.com/timescale/promscale/blob/99c564a8f5da7b8a2c2bcebfa87215d4afdbd81c/pkg/migrations/sql/idempotent/base.sql#L1591-L1596), it seems that that's just a wrapper around [this](https://github.com/timescale/promscale/blob/99c564a8f5da7b8a2c2bcebfa87215d4afdbd81c/pkg/migrations/sql/idempotent/base.sql#L1579-L1589).

So if I understand correctly, Promscale has created two instances of the same job and the docs recommend creating a third instance of the job. I'm speculating but suspect that the jobs are stepping on each other's toes.

In addition, I see lots of messages like this in my Postgres logs:

```
Apr 20 08:28:32 Machine postgres[1011526]: 2021-04-20 08:28:32.547 CEST [1011526] CONTEXT:  SQL statement "                                                                                                                                              
Apr 20 08:28:32 Machine postgres[1011526]:             WITH confirmed_drop_labels AS (                                                                                                                                                                   
Apr 20 08:28:32 Machine postgres[1011526]:                     SELECT label_id                                                                                                                                                                           
Apr 20 08:28:32 Machine postgres[1011526]:                     FROM unnest($1) as labels(label_id)                                                                                                                                                       
Apr 20 08:28:32 Machine postgres[1011526]:                     WHERE NOT EXISTS (                                                                                                                                                                        
Apr 20 08:28:32 Machine postgres[1011526]:                         SELECT 1                                                                                                                                                                              
Apr 20 08:28:32 Machine postgres[1011526]:                         FROM  _prom_catalog.series series_exists                                                                                                                                              
Apr 20 08:28:32 Machine postgres[1011526]:                         WHERE series_exists.labels && ARRAY[labels.label_id]                                                                                                                                  
Apr 20 08:28:32 Machine postgres[1011526]:                         LIMIT 1                                                                                                                                                                               
Apr 20 08:28:32 Machine postgres[1011526]:                     )                                                                                                                                                                                         
Apr 20 08:28:32 Machine postgres[1011526]:                 )                                                                                                                                                                                             
Apr 20 08:28:32 Machine postgres[1011526]:                 DELETE FROM _prom_catalog.label                                                                                                                                                               
Apr 20 08:28:32 Machine postgres[1011526]:                 WHERE id IN (SELECT * FROM confirmed_drop_labels) AND key != '__name__';                                                                                                                      
Apr 20 08:28:32 Machine postgres[1011526]:             "                                                                                                                                                                                                 
Apr 20 08:28:32 Machine postgres[1011526]:         PL/pgSQL function delete_expired_series(text,timestamp with time zone) line 48 at EXECUTE                                                                                                             
Apr 20 08:28:32 Machine postgres[1011526]:         SQL statement "SELECT _prom_catalog.delete_expired_series(metric_table, ran_at)"                                                                                                                      
Apr 20 08:28:32 Machine postgres[1011526]:         PL/pgSQL function drop_metric_chunks(text,timestamp with time zone,timestamp with time zone) line 76 at PERFORM                                                                                       
Apr 20 08:28:32 Machine postgres[1011526]:         SQL statement "CALL _prom_catalog.drop_metric_chunks(r.metric_name, NOW() - _prom_catalog.get_metric_retention_period(r.metric_name))"                                                                
Apr 20 08:28:32 Machine postgres[1011526]:         PL/pgSQL function execute_data_retention_policy() line 16 at CALL                                                                                                                                     
Apr 20 08:28:32 Machine postgres[1011526]:         SQL statement "CALL _prom_catalog.execute_data_retention_policy()"                                                                                                                                    
Apr 20 08:28:32 Machine postgres[1011526]:         PL/pgSQL function execute_maintenance() line 3 at CALL                                                                                                                                                
Apr 20 08:28:32 Machine postgres[1011526]:         SQL statement "CALL prom_api.execute_maintenance()"                                                                                                                                                   
Apr 20 08:28:32 Machine postgres[1011526]:         PL/pgSQL function execute_maintenance_job(integer,jsonb) line 3 at CALL
```

So I suspect this issue might be related to the fact that I have 8630 labels in my database. If you're wondering why, most of them come from identifiers.

**Comment by bobobo1618** [comment]
I'll add that I also have 1121 hypertables.

**Comment by cevian** [comment]
@bobobo1618 What is the rest of the log line in the logs. I see the CONTEXT
`Apr 20 08:28:32 Machine postgres[1011526]: 2021-04-20 08:28:32.547 CEST [1011526] CONTEXT:  SQL statement`

but not the message above it (what is it reporting). Thanks.

**Comment by bobobo1618** [comment]
Sorry, here you go:

```
Apr 20 08:28:32 Machine postgres[1011526]: 2021-04-20 08:28:32.547 CEST [1011526] LOG:  job 1000 threw an error
Apr 20 08:28:32 Machine postgres[1011526]: 2021-04-20 08:28:32.547 CEST [1011526] ERROR:  out of shared memory
Apr 20 08:28:32 Machine postgres[1011526]: 2021-04-20 08:28:32.547 CEST [1011526] HINT:  You might need to increase max_locks_per_transaction.
Apr 20 08:28:32 Machine postgres[1011526]: 2021-04-20 08:28:32.547 CEST [1011526] CONTEXT:  SQL statement "
Apr 20 08:28:32 Machine postgres[1011526]:             WITH confirmed_drop_labels AS (
Apr 20 08:28:32 Machine postgres[1011526]:                     SELECT label_id
Apr 20 08:28:32 Machine postgres[1011526]:                     FROM unnest($1) as labels(label_id)
Apr 20 08:28:32 Machine postgres[1011526]:                     WHERE NOT EXISTS (
Apr 20 08:28:32 Machine postgres[1011526]:                         SELECT 1
Apr 20 08:28:32 Machine postgres[1011526]:                         FROM  _prom_catalog.series series_exists
Apr 20 08:28:32 Machine postgres[1011526]:                         WHERE series_exists.labels && ARRAY[labels.label_id]
Apr 20 08:28:32 Machine postgres[1011526]:                         LIMIT 1
Apr 20 08:28:32 Machine postgres[1011526]:                     )
Apr 20 08:28:32 Machine postgres[1011526]:                 )
Apr 20 08:28:32 Machine postgres[1011526]:                 DELETE FROM _prom_catalog.label
Apr 20 08:28:32 Machine postgres[1011526]:                 WHERE id IN (SELECT * FROM confirmed_drop_labels) AND key != '__name__';
Apr 20 08:28:32 Machine postgres[1011526]:             "
Apr 20 08:28:32 Machine postgres[1011526]:         PL/pgSQL function delete_expired_series(text,timestamp with time zone) line 48 at EXECUTE
Apr 20 08:28:32 Machine postgres[1011526]:         SQL statement "SELECT _prom_catalog.delete_expired_series(metric_table, ran_at)"
Apr 20 08:28:32 Machine postgres[1011526]:         PL/pgSQL function drop_metric_chunks(text,timestamp with time zone,timestamp with time zone) line 76 at PERFORM
Apr 20 08:28:32 Machine postgres[1011526]:         SQL statement "CALL _prom_catalog.drop_metric_chunks(r.metric_name, NOW() - _prom_catalog.get_metric_retention_period(r.metric_name))"
Apr 20 08:28:32 Machine postgres[1011526]:         PL/pgSQL function execute_data_retention_policy() line 16 at CALL
Apr 20 08:28:32 Machine postgres[1011526]:         SQL statement "CALL _prom_catalog.execute_data_retention_policy()"
Apr 20 08:28:32 Machine postgres[1011526]:         PL/pgSQL function execute_maintenance() line 3 at CALL
Apr 20 08:28:32 Machine postgres[1011526]:         SQL statement "CALL prom_api.execute_maintenance()"
Apr 20 08:28:32 Machine postgres[1011526]:         PL/pgSQL function execute_maintenance_job(integer,jsonb) line 3 at CALL
Apr 20 08:28:32 Machine postgres[3181504]: 2021-04-20 08:28:32.565 CEST [3181504] LOG:  background worker "User-Defined Action [1000]" (PID 1011526) exited with exit code 1
```

**Comment by cevian** [comment]
Still need to look into the query. But, have you tried increasing the max_locks_per_transaction PostgreSQL parameter?

**Comment by cevian** [comment]
Looking over this, I suspect that the default of  max_locks_per_transaction = 64 is just too low for 1121 hypertables. What may be happening with the jobs is that they do a lot of work, then reach this query (which is the final one), run out of locks, fail, and undo all the work they just did. It would be interesting to know if the situation improves with max_locks_per_transaction increase.

**Comment by bobobo1618** [comment]
I updated my max_locks_per_transaction to 512 (and ran `timescaledb-tune`) and `CALL prom_api.execute_maintenance();` now completes without throwing errors but takes ~45 minutes to complete, pegging an entire CPU core the entire time.

However after having not run it at all for a couple of weeks, it only took 85 mins to complete, so I think it's going to make the most sense for me to just run it less frequently.

But I don't think that what I'm doing should be super unusual. My "hypertables" are just metrics from a Prometheus instance that were migrated over. So perhaps the default frequency should be reduced and a recommendation to bump up `max_locks_per_transaction` should be documented?

**Comment by Harkishen-Singh** [comment]
We should definitely document this.

> However after having not run it at all for a couple of weeks

We recommend running cron job for this every 30 minutes for best results.

**Comment by bobobo1618** [comment]
> We recommend running cron job for this every 30 minutes for best results.

Upon what is that recommendation based?

Right now, every 30 minutes isn't an option, because the job doesn't complete within 30 minutes. I can either:

- Run it as frequently as possible, in which case it costs 1 CPU core. 
- Run it once an hour, in which case it takes 45 minutes and costs 0.75 CPU cores. 
- Run it once a day in which case I expect it to take around 50 minutes, costing 0.035 CPU cores. 
- Run it once every couple of weeks, in which case it takes around 80 minutes, costing next to nothing in terms of CPU. 

There's clearly a tradeoff to be made here but what's the benefit I should expect to gain from the recommendation to pin a CPU core instead of running the task less frequently?

**Comment by Harkishen-Singh** [comment]
I think the reason why its taking that much time is you have uncompressed chunks that were pending from long. 

When execute_maintenance is executed, it goes through all metric hypertables and checks if there are any pending chunks that can be compressed. If yes, it executes compression on them, otherwise moves to the next metric.

I think the math you are doing is averaging the CPU over time which is fine, but then you get a lot more chunks to be compressed and hence takes high CPU and time. If you run it every 30 mins, you should not face the high CPU issue after a, few trys, since the pending chunks should fall decrease.

I think the best way here would be to try once executing every 30 mins for a while and see if it takes the same time or improves with successive tries.

**Comment by bobobo1618** [comment]
> I think the reason why its taking that much time is you have uncompressed chunks that were pending from long.

But when I run it again _immediately after it finishes_, it still takes 45 minutes. 45 minutes is the minimum amount of time it takes to complete.

> I think the best way here would be to try once executing every 30 mins for a while

I _can't_ run it every 30 minutes because it takes longer than 30 minutes to complete.

**Comment by Harkishen-Singh** [comment]
> But when I run it again immediately after it finishes, it still takes 45 minutes.

Got it.

Can you just run `CALL _prom_catalog.execute_compression_policy()` successively and see how much time it takes?

**Comment by bobobo1618** [comment]
That on its own consistently takes ~31 minutes.

**Comment by Harkishen-Singh** [comment]
Then I guess the time taken for the checks is significant. In that case, should be good to go with the math you mentioned above, regarding per ~week.

**Comment by cevian** [comment]
I believe this may be fixed by #678

**Comment by VineethReddy02** [comment]
@cevian can we close this issue?
